Core/include/Acts/EventData/detail/residual_calculator.hpp

```diff
@@ -24,43 +25,59 @@ namespace detail {
 ///
 /// @return residual_calculator<params...>::result(first,second) yields the
 ///         residuals of `first` with respect to `second`
-template <ParID_t... params>
+template <typename parameter_indices_t, unsigned int... params>
 struct residual_calculator;
-
 /// @cond
-template <typename R, ParID_t... params>
+
+template <typename parameter_indices_t, typename R, unsigned int... params>
 struct residual_calculator_impl;
 
-template <ParID_t... params>
+template <typename parameter_indices_t, unsigned int... params>
 struct residual_calculator {
   using ParVector_t = ActsVector<ParValue_t, sizeof...(params)>;
 
   static ParVector_t result(const ParVector_t& test, const ParVector_t& ref) {
     ParVector_t result;
-    residual_calculator_impl<ParVector_t, params...>::calculate(result, test,
-                                                                ref, 0);
+    residual_calculator_impl<parameter_indices_t, ParVector_t,
+                             params...>::calculate(result, test, ref, 0);
     return result;
   }
 };
 
-template <typename R, ParID_t first, ParID_t... others>
-struct residual_calculator_impl<R, first, others...> {
+template <typename parameter_indices_t, typename R, unsigned int first,
+          unsigned int... others>
+struct residual_calculator_impl<parameter_indices_t, R, first, others...> {
   static void calculate(R& result, const R& test, const R& ref,
                         unsigned int pos) {
-    result(pos) = BoundParameterType<first>::getDifference(test(pos), ref(pos));
-    residual_calculator_impl<R, others...>::calculate(result, test, ref,
-                                                      pos + 1);
+    if constexpr (std::is_same<parameter_indices_t,
+                               BoundParametersIndices>::value) {
+      result(pos) = BoundParameterType<static_cast<BoundParametersIndices>(
+          first)>::getDifference(test(pos), ref(pos));
+    } else {
+      result(pos) = FreeParameterType<static_cast<FreeParametersIndices>(
+          first)>::getDifference(test(pos), ref(pos));
+    }
```

Comment:
This here is a bit ugly, because you need to map from the indices type to the params type. Could we maybe pull that out and use specialization like for `FreeParameterTraits` to have that mapping next to where `BoundParameterType` is defined. Something like 

```cpp
template <typename T>
struct ParamTypeFor {}:

template <>
struct ParamTypeFor<BoundParametersIndices> {
  using type = BoundParameterType;
};
```

You can probably

Reply:
The `BoundParameterType` is needs the index as template parameter, so I can offer you a struct templated on the type and the index.

Reply:
Changed it